### PR TITLE
crimson/osd/osd_operation: fix dump_historic_slow_ops command works

### PR DESCRIPTION
**Pull request by guojidan**
Fixes: https://tracker.ceph.com/issues/65531

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

**Comment by guojidan** [comment]
jenkins test make check

**Comment by guojidan** [comment]
jenkins test windows

**Comment by guojidan** [comment]
jenkins test make check arm64

**Comment by athanatos** [comment]
The commit message "crimson/osd/osd_operation.cc: optimize the calculation and output of slow_ops" suggests that this is a performance optimization.  In fact, from the linked bug, this actually fixes a few bugs in how the command works.  Please update the commit message to explain what this commit changes and how it fixes the bugs.

**Comment by athanatos** [comment]
As a guideline, "optimize ..." is basically never an acceptable commit message as it really doesn't tell a future reader what the commit is changing.

File: src/crimson/osd/osd_operation.cc [hunk 1]
```diff
@@ -63,7 +63,7 @@ void OSDOperationRegistry::put_historic(const ClientRequest& op)
     static_cast<size_t>(OperationTypeCode::historic_client_request);
   auto& client_registry = get_registry<client_reg_index>();
   auto& historic_registry = get_registry<historic_reg_index>();
-  historic_registry.splice(std::end(historic_registry),
+  historic_registry.splice(last_of_recents,
```
Comment (athanatos) [inline src/crimson/osd/osd_operation.cc:66]: Don't we always want to insert at the end of historic_registry?

File: src/crimson/osd/osd_operation.cc [hunk 2]
```diff
@@ -75,11 +75,11 @@ void OSDOperationRegistry::put_historic(const ClientRequest& op)
   // NOTE: Operation uses the auto-unlink feature of boost::intrusive.
   // NOTE: the cleaning happens in OSDOperationRegistry::do_stop()
   using crimson::common::local_conf;
-  if (num_recent_ops >= local_conf()->osd_op_history_size) {
-    ++last_of_recents;
+  if (get_duration(op) > local_conf()->osd_op_complaint_time) {
     ++num_slow_ops;
   } else {
     ++num_recent_ops;
+    --last_of_recents;
```
Comment (athanatos) [inline src/crimson/osd/osd_operation.cc:82]: I'm confused as to what last_of_recents is supposed to indicate.  Is it the case that for all ops in the interval [last_of_recents, std::end(historic_registry)), get_duration(op) <= local_conf()->osd_op_complaint_time?  If so, I'm not sure this patch actually maintains that invariant.

File: src/crimson/osd/osd_operation.cc [hunk 3]
```diff
@@ -100,6 +100,10 @@ void OSDOperationRegistry::put_historic(const ClientRequest& op)
     ClientRequest::ICRef(&fastest_historic_op, /* add_ref= */false);
     --num_slow_ops;
   }
+  if (num_recent_ops > local_conf()->osd_op_history_size) {
+    historic_registry.pop_back();
+    --num_recent_ops;
```
Comment (athanatos) [inline src/crimson/osd/osd_operation.cc:105]: The last op may not be recent.

**Review by athanatos — CHANGES_REQUESTED** [review]
See my other comments -- I might just be misunderstanding the patch, though.

**Comment by guojidan** [comment]
hi @athanatos , all historic_client_request ops are in one `op_list`, so I want to separate them, The first `local_conf()->osd_op_history_slow_op_size` elements in the `op_list` store the slow ops log,  while the last `local_conf()->osd_op_history_size` elements  store the normal history ops log, use `last_of_recents` as separated markers.

insert slow ops before `last_of_recents`, normal ops insert at the end of `op_list`

**Comment by athanatos** [comment]
Perhaps it would be simpler to use two separate lists rather than maintain last_of_recents as a separator?

**Comment by guojidan** [comment]
> Perhaps it would be simpler to use two separate lists rather than maintain last_of_recents as a separator?

ok, I will reimplement this

**Review by athanatos — APPROVED** [review]
Ah, that's quite clean.  LGTM!

**Comment by Matan-B** [comment]
https://shaman.ceph.com/builds/ceph/wip-matanb-crimson-testing-june-9/

**Comment by Matan-B** [comment]
jenkins test make check

**Comment by Matan-B** [comment]
Merging based on (21/21 passed):

https://pulpito.ceph.com/matan-2024-06-09_09:28:06-crimson-rados-wip-matanb-crimson-testing-june-9-distro-crimson-smithi/
rerun dead jobs:
https://pulpito.ceph.com/matan-2024-06-09_10:57:06-crimson-rados-wip-matanb-crimson-testing-june-9-distro-crimson-smithi/

**Comment by Matan-B** [comment]
jenkins test make check

**Comment by Matan-B** [comment]
jenkins test make check

**Comment by xxhdx1985126** [comment]
Seems that this PR is caused memory leak within the OSD

**Comment by guojidan** [comment]
> Seems that this PR is causing memory leak within the OSD

can you provide more infomation?

**Comment by xxhdx1985126** [comment]
> > Seems that this PR is causing memory leak within the OSD
> 
> can you provide more infomation?

With the commit in this PR, no client request gets destroyed. If you set debug_osd to 25, all client request logs except "client_request.*destroying" are printed.

**Comment by Matan-B** [comment]
Leak seems intentional though cleanup was omitted here.

```
  ClientRequest::ICRef(
    &op, /* add_ref= */true
  ).detach(); // yes, "leak" it for now!
```
 Is there, while the cleanup is removed
```
    // clear a previously "leaked" op
    ClientRequest::ICRef(&fastest_historic_op, /* add_ref= */false);
```

**Comment by guojidan** [comment]
> Leak seems intentional though cleanup was omitted here.泄漏似乎是故意的，尽管这里省略了清理。
> 
> ```
>   ClientRequest::ICRef(
>     &op, /* add_ref= */true
>   ).detach(); // yes, "leak" it for now!
> ```
> 
> Is there, while the cleanup is removed在那里，而清理被删除
> 
> ```
>     // clear a previously "leaked" op
>     ClientRequest::ICRef(&fastest_historic_op, /* add_ref= */false);
> ```

Oh, I see, I will fix this.

**Comment by Matan-B** [comment]
Revert PR here: https://github.com/ceph/ceph/pull/58223
@guojidan, if the revert PR will be merged before the fix feel free to re-push the entire branch again (with the fix).
Thank you!

**Pull request merged**